services/shhext/dedup/cache.go

```diff
@@ -0,0 +1,111 @@
+package dedup
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto/sha3"
+	whisper "github.com/ethereum/go-ethereum/whisper/whisperv6"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+// cache represents a cache of whisper messages with a limit of 2 days.
+// the limit is counted from the time when the message was added to the cache.
+type cache struct {
+	db  *leveldb.DB
+	now func() time.Time
+}
+
+func newCache(db *leveldb.DB) *cache {
+	return &cache{db, time.Now}
+}
+
+func (d *cache) Has(filterID string, message *whisper.Message) (bool, error) {
+	has, err := d.db.Has(d.keyToday(filterID, message), nil)
+
+	if err != nil {
+		return false, err
+	}
+	if has {
+		return true, nil
+	}
+
+	return d.db.Has(d.keyYesterday(filterID, message), nil)
+}
+
+func (d *cache) Put(filterID string, messages []*whisper.Message) error {
+	batch := leveldb.Batch{}
+
+	for _, msg := range messages {
+		batch.Put(d.keyToday(filterID, msg), []byte{})
+	}
+
+	err := d.db.Write(&batch, nil)
+	if err != nil {
+		return err
+	}
+
+	return d.cleanOldEntries()
+}
+
+func (d *cache) cleanOldEntries() error {
+	// Cleaning up everything that is older than 2 days
+	// We are using the fact that leveldb can do prefix queries and that
+	// the entries are sorted by keys.
+	// Here, we are looking for all the keys that are between
+	// 00000000.* and <yesterday's date>.*
+	// e.g. (0000000.* -> 20180424.*)
+
+	limit := d.yesterdayDateString()
+
+	r := &util.Range{
+		Start: []byte("00000000"),
+		Limit: []byte(limit),
+	}
+
+	batch := leveldb.Batch{}
+	iter := d.db.NewIterator(r, nil)
+	for iter.Next() {
+		batch.Delete(iter.Key())
+	}
+	iter.Release()
+
+	return d.db.Write(&batch, nil)
+}
+
+func (d *cache) keyYesterday(filterID string, message *whisper.Message) []byte {
+	return prefixedKey(d.yesterdayDateString(), filterID, message)
+}
+
+func (d *cache) keyToday(filterID string, message *whisper.Message) []byte {
+	return prefixedKey(d.todayDateString(), filterID, message)
+}
+
+func (d *cache) todayDateString() string {
+	return dateString(d.now())
+}
+
+func (d *cache) yesterdayDateString() string {
+	now := d.now()
+	yesterday := now.Add(-24 * time.Hour)
+	return dateString(yesterday)
+}
+
+func dateString(t time.Time) string {
+	// Layouts must use the reference time Mon Jan 2 15:04:05 MST 2006
+	return t.Format("20060102")
+}
+
+func prefixedKey(date, filterID string, message *whisper.Message) []byte {
+	prefix := fmt.Sprintf("DD-%s-%s", date, filterID)
+	return append([]byte(prefix), key(message)[:]...)
+}
+
+func key(message *whisper.Message) []byte {
```

Comment:
@mandrigin idk why but my answer got lost

> is that a hash of content?

it includes hash of the content and all other envelope fields

Reply:
yeah, that exactly what I don't want 😀 
I want just content, topic and user.

Reply:
so for example, if i will send 3 x `help!` in fast series to emphasize that i really need it - will you just filter it as a duplicates or it will be different content because of some nonce in the message payload?

also, does our payload contain timestamp? i remember that @cammellos was going to reuse whisper message timestamp. if so it could happen that same message will be deduplicated even when it shouldn't

Reply:
each message used to have a unique id... let me check the new protocol though.

Reply:
and right now the payload has a timestamp in nanoseconds, afaik.
but that's not good enough 😀 
though, why would we use Whisper timestamp? how would that work with the messages received from the mailserver? wouldn't that break the time display?

Reply:
@dshulyak from @janherich (paraphrasing)

> That won’t happen because each chat message contains logical clock and this is always increasing (for given sender) and it's a part of the payload

Reply:
thanks that is also what i meant in 
> because of some nonce in the message payload